[Local GC] Remove uses of g_SystemInfo from the GC

swgillespie opened the pull request:
Fixes https://github.com/dotnet/coreclr/issues/11512, part of https://github.com/dotnet/coreclr/issues/11518.

This PR removes uses of `g_SystemInfo` from within the GC. The uses of `g_SystemInfo` fall into three categories:

* Use of `g_SystemInfo.dwNumberOfProcessors` to get the number of processors. These have been replaced with a reference to `g_num_processors`, a global initialized at GC initialization time with the return result of `GCToOSInterface::GetCurrentProcessCpuCount`. Outside of the GC (i.e. the handle table), `g_SystemInfo.dwNumberOfProcessors` is replaced with a direct call to `GCToOSInterface::GetCurrentProcessCpuCount`.
* Use of `g_SystemInfo.dwPageSize` to get the OS page size. These are replaced with the `OS_PAGE_SIZE` macro.
* Use of `g_SystemInfo.dwAllocationGranularity`. When reserving the card table, the GC manually aligns the reservation size up to the OS allocation granularity. I did some reading and I don't think this is strictly necessary; Windows will do the alignment itself if necessary and the pointer it returns is always aligned to the allocation granularity. I found that I still had to align to the allocation granularity for the PAL implementation of `GCToOSInterface`, since `EEVirtualAlloc` [requires that mappings without a hint address be allocation granularity aligned](https://github.com/dotnet/coreclr/blob/master/src/vm/hosting.cpp#L220). At any rate, my solution in this PR was to remove the GC's manual alignment and rely instead on the OS's natural alignment.

I'm in the process of validating these changes.

swgillespie commented:
@dotnet-bot test Windows_NT x64 Checked GC Reliability Framework

swgillespie commented:
@dotnet-bot test Windows_NT Checked gc_reliability_framework

swgillespie commented:
@Maoni0 @adityamandaleeka @sergiy-k @jkotas PTAL? @janvorli you may also be interested? I'm least sure about the page granularity alignment changes when reserving memory.

Maoni0 commented on src/gc/gc.cpp:
```diff
@@ -7417,7 +7414,7 @@ int gc_heap::grow_brick_card_tables (uint8_t* start,
 
             //delete (uint32_t*)((uint8_t*)ct - sizeof(card_table_info));
-            if (!GCToOSInterface::VirtualRelease (mem, alloc_size_aligned))
+            if (!GCToOSInterface::VirtualRelease (mem, alloc_size))
```
>alloc_size [](start = 55, length = 10)

this arg is actually not used, could you please remove it from the VirtualRelease call's signature?

jkotas replied:
The argument is used on Unix

Maoni0 replied:
oh okie

Maoni0 commented on src/gc/gc.cpp:
```diff
@@ -7035,7 +7033,7 @@ uint32_t* gc_heap::make_card_table (uint8_t* start, uint8_t* end)
     card_table_lowest_address (ct) = start;
     card_table_highest_address (ct) = end;
     card_table_brick_table (ct) = (short*)((uint8_t*)ct + cs);
-    card_table_size (ct) = alloc_size_aligned;
```
>card_table_size [](start = 4, length = 15)

if you remove the size from the VirtualRelease call, card_table_size can also be removed.

Maoni0 replied:
then ignore this comment :P

Maoni0 commented on src/gc/gc.cpp:
```diff
@@ -33623,6 +33620,8 @@ HRESULT GCHeap::Initialize ()
     }
 
     g_gc_pFreeObjectMethodTable = GCToEEInterface::GetFreeObjectMethodTable();
+    g_num_processors = GCToOSInterface::GetCurrentProcessCpuCount();
```
>GetCurrentProcessCpuCount [](start = 40, length = 25)

this should not be obtained from GetCurrentProcessCpuCount which gives you how many CPUs this process is affinitized to. You could be on a 64-CPU system and the process could be affinitized to 2 of them. This should be for the whole machine (ie, in win32 it should be from GetSystemInfo).

swgillespie replied:
Ok! In that case I'm not sure if there's an existing API on `GCToOSInterface` to do that. `GCToOSInterface::GetLogicalCpuCount` ends up calling [this](https://github.com/dotnet/coreclr/blob/master/src/vm/amd64/cgenamd64.cpp#L464). I remember us receiving some guidance from Intel about updating this function - do you think now would be a good time for that?

Maoni0 replied:
GetLogicalCpuCount is for something different - that's for getting how many logical CPUs are on one physical processor. You could just add a new API for getting the # of processors on the machine.

swgillespie replied:
Will do!

swgillespie commented:
My local run of the RF with a standalone GC just crashed and burned with https://github.com/dotnet/coreclr/issues/11282, so the RF run on this PR might be doomed...

janvorli commented on src/gc/handletablecore.cpp:
```diff
@@ -516,14 +516,14 @@ BOOL SegmentInitialize(TableSegment *pSegment, HandleTable *pTable)
 #ifndef FEATURE_REDHAWK // todo: implement SafeInt
     // Prefast overflow sanity check the addition
-    if (!ClrSafeInt<uint32_t>::addition(dwCommit, g_SystemInfo.dwPageSize, dwCommit))
+    if (!ClrSafeInt<uint32_t>::addition(dwCommit, OS_PAGE_SIZE, dwCommit))
```
This needs to be the real OS size determined at runtime since that's the granularity with which the memory can be committed. For example, the ARM64 Linux distros have varying page size.
The same is true for all other places below where you have replaced the g_SystemInfo.dwPageSize with OS_PAGE_SIZE.

swgillespie replied:
Ok! In that case, what is the purpose of `OS_PAGE_SIZE`? Is that only for platforms whose page sizes are constant?

janvorli replied:
Ah, I have not realized that the change to support larger page sizes on Unix is not in yet and that that change will solve that (it defines OS_PAGE_SIZE as a call to a new GCToOSInterface::GetPageSize()). So please ignore my comment.

swgillespie replied:
Oh, ok! In that case, @sdmaclea does this look OK to you?

sdmaclea replied:
Yes that looks OK

swgillespie commented:
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test

swgillespie commented:
Since there haven't been any objections and I believe I've addressed all feedback I'm going to go ahead and merge this into the feature branch. It'll be flowing into master as a part of https://github.com/dotnet/coreclr/pull/11699.
